Firestore/core/src/firebase/firestore/remote/serializer.h

```diff
@@ -111,7 +116,56 @@ class Serializer {
   firebase::firestore::model::DocumentKey DecodeKey(
       absl::string_view name) const;
 
+  /**
+   * @brief Converts the Document (i.e. key/value) into bytes.
+   *
+   * @param[out] out_bytes A buffer to place the output. The bytes will be
+   * appended to this vector.
+   * @return A Status, which if not ok(), indicates what went wrong. Note that
+   * errors during encoding generally indicate a serious/fatal error.
+   */
+  // TODO(rsgowman): Similar to above, if we never support any output except to
+  // a vector, it may make sense to have Serializer own the vector and provide
+  // an accessor rather than asking the user to create it first.
+  util::Status EncodeDocument(
```

Comment:
There's an asymmetry here. We're not *quite* encoding a Document, we're only encoding the relevant subset of a document. (We ignore create/update time.)

Perhaps a more natural way to do this, would be to treat the (lack of) encoding of create/update time as an implementation detail, and just pass in the entire Document, rather than just the fields to be encoded (though it makes it less obvious that those fields aren't being encoded.) However, I've encoded it this way to mirror the java port.

Reply:
> Perhaps a more natural way to do this, would be to treat the (lack of) encoding of create/update time as an implementation detail

This sounds more natural to me (though I might be missing some context), but probably not enough to justify divergence from other platforms (assuming Android and Web agree here).

Reply:
For now, I'll leave it as is. (Consistency's a pretty strong reason to leave as is.) If it starts to annoy me, I'll propose something against all ports. Plus, we'll get a chance to revisit during the local serializer implementation.

> IIRC Documents containing them are rejected. 

That seems a little harsh, especially since (presumably) documents containing unknown proto fields wouldn't be. But I suppose it does drive home the fact that they'll otherwise be ignored. :)